File: ideaproject/javaserializer/src/com/houseelectrics/serializer/DefaultFinder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace com.houseelectrics.serializer
{

    public class DefaultFinder
    {
        //, TypeAliaser aliaser
        public LeafDefaultSet getDefaultsForAllLinkedObjects(Object root, NodeExpander nodeExpander)
        {
            LeafDefaultSet result = new LeafDefaultSet();

            ObjectExplorerImpl explorer = new ObjectExplorerImpl();
                MoveAway down = ( from,  propertyName,  to,  isIndexed, index) =>
                {
                    return (isIndexed || (to != null && !result.Type2Defaults.ContainsKey(to.GetType())));
                };
             MoveBack  up = (from, propertyName, to, isIndexed) =>
                {

                };

             OnLeaf leaf = (from, propertyName, to, index) =>
                {
                    Defaults4Class defaults = null;
                    if (from==null) {return;}
                    Type fromType = from.GetType();
                    if (!result.Type2Defaults.ContainsKey(fromType))
                    {
                        defaults = new Defaults4Class();
                        defaults.FullClassName = fromType.FullName;
                        result.Type2Defaults[fromType] = defaults;
                    }
                    else
                    {
                        defaults = result.Type2Defaults[fromType];
                    }
                    defaults.PropertyName2DefaultValue[propertyName] = to;
                };

             explorer.NodeExpander = nodeExpander;
                explorer.explore(root, down, up, leaf);
                return result;
        }
    }
}

File: ideaproject/javaserializer/src/com/houseelectrics/serializer/DefaultFinder.java
package com.houseelectrics.serializer;

/**
 * Created by roberttodd on 01/12/2014.
 */
public class DefaultFinder
{
    public LeafDefaultSet getDefaultsForAllLinkedObjects(Object root, NodeExpander nodeExpander)
    {
        final LeafDefaultSet result = new LeafDefaultSet();

        ObjectExplorerImpl explorer = new ObjectExplorerImpl();

        ExplorationListener listener = new ExplorationListener()
        {
            @Override
            public boolean MoveAway(Object from, String propertyName, Object to, boolean isIndexed, Integer index)
            {
                return (isIndexed || (to != null && !result.getType2Defaults().containsKey(to.getClass())));
            }

            @Override
            public void MoveBack(Object from, String propertyName, Object to, boolean isIndexed)
            {
            }

            @Override
            public void OnLeaf(Object from, String propertyName, Object to, Integer index)
            {
                Defaults4Class defaults = null;
                if (from==null) {return;}
                Class fromType = from.getClass();
                if (!result.getType2Defaults().containsKey(fromType))
                {
                    defaults = new Defaults4Class();
                    defaults.setFullClassName(fromType.getName());
                    result.getType2Defaults().put(fromType, defaults);
                }
                else
                {
                    defaults = result.getType2Defaults().get(fromType);
                }
                defaults.getPropertyName2DefaultValue().put(propertyName, to);
            }
        };

        explorer.setNodeExpander(nodeExpander);
        explorer.explore(root, listener);
        return result;
    }
}
